src/dragonball/src/tracer.rs

```diff
@@ -0,0 +1,287 @@
+        {
+            return true;
+        }
+        false
+    }
+}
+
+impl Eq for TraceConfigInfo {}
+
+/// drabonball tracer.
+/// The dragonball tracer can run in two situations:
+///     1. dragonball runs as lib, such as running in kata.
+///     2. dragonball runs as app, such as running with `dbs-cli`.
+pub struct DragonballTracer {
+    /// Dragonball tracing subscriber.
+    subscriber: Arc<dyn Subscriber + Send + Sync>,
+    /// Subscriber guard. The lifetime is as same as tracer.
+    sub_guard: Option<DefaultGuard>,
+    /// The subscriber obtained from runtime.
+    from_runtime: bool,
+    /// Enable tracing
+    enable_tracing: bool,
+    /// Tracing's async runtime and lifetime is as same as tracer.
+    tokio_rt: Runtime,
+    /// Dragonball tracing root span.
+    root_span: Option<Arc<EnteredSpan>>,
+}
+
+impl Default for DragonballTracer {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl DragonballTracer {
+    /// Create a dragonball tracer instance.
+    pub fn new() -> Self {
+        Self {
+            subscriber: Arc::new(NoSubscriber::default()),
+            sub_guard: None,
+            from_runtime: false,
+            enable_tracing: false,
+            tokio_rt: tokio::runtime::Builder::new_multi_thread()
+                .enable_all()
+                .build()
+                .expect("Tracer: tokio new thread err"),
+            root_span: None,
+        }
+    }
+
+    /// Return whether the tracing is enabled, enabled by [`setup_tracing`]
+    pub fn enabled(&self) -> bool {
+        self.enable_tracing
+    }
+
+    /// tracing feature for app.
+    pub fn setup_tracing(&mut self, config: TraceConfigInfo) -> Result<(), DragonballTraceError> {
+        if !self.enable_tracing {
+            self.enable_tracing = true;
+            // tracing in application.
+            if config.subscriber.is_none() {
+                let mut config = config;
+                Self::verify_jaeger_config(&mut config)
+                    .map_err(DragonballTraceError::ConfigVerifyFailed)?;
+
+                let subscriber: Arc<Mutex<Arc<dyn Subscriber + Send + Sync>>> =
+                    Arc::new(Mutex::new(Arc::new(NoSubscriber::default())));
+                let shared_sub = subscriber.clone();
+                self.tokio_rt.block_on(async {
+                    let tracer = opentelemetry_jaeger::new_collector_pipeline()
+                        .with_service_name(format!(
+                            "dragonball-{}",
+                            &config.sid[0..min(8, config.sid.len())]
+                        ))
+                        .with_endpoint(config.jaeger_endpoint)
+                        .with_username(config.jaeger_user)
+                        .with_password(config.jaeger_password)
+                        .with_hyper()
+                        .install_batch(opentelemetry::runtime::Tokio)
+                        .expect("Tracer: create jaeger tracer err");
```

Comment:
return error instead of expect

Reply:
This code is running in the closure of `block_on` function, we can't process the error outside the closure. I add a comment here.